gradio/examples.py

```diff
@@ -0,0 +1,191 @@
+        examples_per_page (int): how many examples to show per page (this parameter currently has no effect)
+        """
+        if not isinstance(inputs, list):
+            inputs = [inputs]
+        if not isinstance(outputs, list):
+            outputs = [outputs]
+
+        if examples is None or (
+            isinstance(examples, list)
+            and (len(examples) == 0 or isinstance(examples[0], list))
+        ):
+            pass
+        elif (
+            isinstance(examples, list) and len(inputs) == 1
+        ):  # If there is only one input component, examples can be provided as a regular list instead of a list of lists
+            examples = [[e] for e in examples]
+        elif isinstance(examples, str):
+            if not os.path.exists(examples):
+                raise FileNotFoundError(
+                    "Could not find examples directory: " + examples
+                )
+            log_file = os.path.join(examples, "log.csv")
+            if not os.path.exists(log_file):
+                if len(inputs) == 1:
+                    exampleset = [
+                        [os.path.join(examples, item)] for item in os.listdir(examples)
+                    ]
+                else:
+                    raise FileNotFoundError(
+                        "Could not find log file (required for multiple inputs): "
+                        + log_file
+                    )
+            else:
+                with open(log_file) as logs:
+                    exampleset = list(csv.reader(logs))
+                    exampleset = exampleset[1:]  # remove header
+            for i, example in enumerate(exampleset):
+                for j, (component, cell) in enumerate(
+                    zip(
+                        inputs + outputs,
+                        example,
+                    )
+                ):
+                    exampleset[i][j] = component.restore_flagged(
+                        examples,
+                        cell,
+                        None,
+                    )
+            examples = exampleset
+        else:
+            raise ValueError(
+                "Examples argument must either be a directory or a nested "
+                "list, where each sublist represents a set of inputs."
+            )
+
+        dataset = Dataset(
+            components=inputs,
+            samples=examples,
+            type="index",
+        )
+
+        self.examples = examples
+        self.inputs = inputs
+        self.outputs = outputs
+        self.fn = fn
+        self.cache_examples = cache_examples
+        self.examples_per_page = examples_per_page
+        self.cached_folder = os.path.join(CACHED_FOLDER, str(dataset._id))
+        self.cached_file = os.path.join(self.cached_folder, "log.csv")
+
+        if cache_examples:
+            self.cache_interface_examples()
+
+        def load_example(example_id):
+            processed_examples = [
```

Comment:
The examples don't change during the lifetime of the app. I think we can compute `processed_examples` once and avoid a bunch of repetitive work?